docs(dockerfile): merge buildkit syntax with reference docs (#2934)
moby/buildkit: frontend/dockerfile/docs/reference.md

  If your system doesn't have support for `dirperm1`, the issue describes a workaround.

## RUN --mount

To use the `--mount` flag in `RUN` command, set Dockerfile version to at least
[`1.2`](#syntax):

> ```dockerfile
> # syntax=docker/dockerfile:1.4
> ```

Reviewer: With the above, perhaps we can remove this line (for the same reason

```suggestion
```

`RUN --mount` allows you to create mounts that process running as part of the
build can access. This can be used to bind files from other part of the build
without copying, accessing build secrets or ssh-agent sockets, or creating cache
locations to speed up your build.

### RUN --mount=type=bind

> **Note**
>

> > `type=bind` is the default mount type being used if not provided

Reviewer: Perhaps we should add a table under `RUN --mount` with the available types, and a "default", something like:

```markdown
Syntax:

    --mount=[type=<TYPE>][,option=<value>[,option=<value>]...]

### Mount types

| Type                                     | Description                                                                                               |
|------------------------------------------|-----------------------------------------------------------------------------------------------------------|
| [`bind`](#run---mounttypebind) (default) | Bind-mount context directories (read-only).                                                               |
| [`cache`](#run---mounttypecache)         | Mount a temporary directory for to cache directories for compilers and package managers..                 |
| [`secret`](#run---mounttypesecret)       | Allow the build container to access SSH keys via SSH agents, with support for passphrases.                |
| [`ssh`](#run---mounttypessh)             | Allow the build container to access secure files such as private keys without baking them into the image. |
```

| `gid`               | Group ID for socket. Default `0`.                                                              |

#### Example: access to Gitlab

```dockerfile
# syntax=docker/dockerfile:1.4
FROM alpine
RUN apk add --no-cache openssh-client
RUN mkdir -p -m 0700 ~/.ssh && ssh-keyscan gitlab.com >> ~/.ssh/known_hosts
RUN --mount=type=ssh \
  ssh -q -T git@gitlab.com 2>&1 | tee /hello
# "Welcome to GitLab, @GITLAB_USERNAME_ASSOCIATED_WITH_SSHKEY" should be printed here
# with the type of build progress is defined as `plain`.
```

```console
$ eval $(ssh-agent)
$ ssh-add ~/.ssh/id_rsa
(Input your passphrase here)
$ docker buildx build --ssh default=$SSH_AUTH_SOCK .
```

You can also specify a path to `*.pem` file on the host directly instead of `$SSH_AUTH_SOCK`.
However, pem files with passphrases are not supported.

## RUN --network

```dockerfile
# syntax=docker/dockerfile:1.4
```

`RUN --network` allows control over which networking environment the command
is run in.

### RUN --network=none

The command is run with no network access (`lo` is still available, but is

> isolated to this process)

Author: I think the title of the section is self-explanatory no?

#### Example: isolating external effects

```dockerfile
# syntax=docker/dockerfile:1.3
FROM python:3.6
ADD mypackage.tgz wheels/
RUN --network=none pip install --find-links wheels mypackage
```

`pip` will only be able to install the packages provided in the tarfile, which
can be controlled by an earlier build stage.

### RUN --network=host

The command is run in the host's network environment (similar to
`docker build --network=host`, but on a per-instruction basis)

> **Warning**
>

> > The use of `--network=host` is protected by the `network.host` entitlement, 

Author: I don't see any docs for entitlements in the repo. I think we should add and md file in `/docs` in a follow-up and update the reference docs for cross linking. Current available entitlements are: https://github.com/moby/buildkit/blob/b57ada1d548a942957dcc4269bb9d93f3e6aeeb2/util/entitlements/entitlements.go#L14-L17

> > The use of `--network=host` is protected by the `network.host` entitlement, 

Author: Oh we can link to https://github.com/docker/buildx/blob/master/docs/reference/buildx_build.md#allow actually.

| [`host`](#run---networkhost)                 | Run in the host's network environment. |

### RUN --network=default

Equivalent to not supplying a flag at all, the command is run in the default
network for the build.

### RUN --network=none

The command is run with no network access (`lo` is still available, but is
isolated to this process)

#### Example: isolating external effects

```dockerfile
# syntax=docker/dockerfile:1
FROM python:3.6
ADD mypackage.tgz wheels/
RUN --network=none pip install --find-links wheels mypackage
```

`pip` will only be able to install the packages provided in the tarfile, which
can be controlled by an earlier build stage.

### RUN --network=host

The command is run in the host's network environment (similar to
`docker build --network=host`, but on a per-instruction basis)

> **Warning**
>
> The use of `--network=host` is protected by the `network.host` entitlement, 
> which needs to be enabled when starting the buildkitd daemon with
> `--allow-insecure-entitlement network.host` flag or in [buildkitd config](https://github.com/moby/buildkit/blob/master/docs/buildkitd.toml.md),

> > and for a build request with [`--allow network.host` flag](https://docs.docker.com/engine/reference/commandline/buildx_build/#allow).

Reviewer 2: Could we import some of the buildx docs into this repo as a follow-up? Seems strange to me that we don't document entitlements in buildkit.

{:.warning}

## RUN --security

> **Note**
>

> > Not yet available in stable syntax, use [labs](#syntax) version.

Reviewer 2: Can we specify an exact labs version, similar to how we do above? Or at least somehow indicate which labs version it's been present in from.
```suggestion
> Not yet available in stable syntax, use [`docker/dockerfile:1.4-labs`](#syntax) version.
```

> following instructions from the Dockerfile if the contents of `<src>` have
> changed. This includes invalidating the cache for `RUN` instructions.
> See the [`Dockerfile` Best Practices
guide – Leverage build cache](https://docs.docker.com/develop/develop-images/dockerfile_best-practices/#leverage-build-cache)
> for more information.

## COPY --link

> **Note**
>
> Added in [`docker/dockerfile:1.4`](#syntax)

Enabling this flag in `COPY` or `ADD` commands allows you to copy files with
enhanced semantics where your files remain independent on their own layer and
don't get invalidated when commands on previous layers are changed.

When `--link` is used your source files are copied into an empty destination
directory. That directory is turned into a layer that is linked on top of your
previous state.

```dockerfile
# syntax=docker/dockerfile:1
FROM alpine
COPY --link /foo /bar
```

Is equivalent of doing two builds:

```dockerfile
FROM alpine
```

and

```dockerfile
FROM scratch
COPY /foo /bar
```

and merging all the layers of both images together.

### Benefits of using `--link`

> Using `--link` allows to reuse already built layers in subsequent builds with

Reviewer 2: ```suggestion
Using `--link` allows reusing already built layers in subsequent builds with
```

Reviewer: Perhaps active voice for these;

```markdown
Use `--link` to reuse already built layers in subsequent builds ...
```

RUN echo "I'm building for $TARGETPLATFORM"
```

### BuildKit built-in build args

| Arg                                   | Type    | Description                                                              |
|---------------------------------------|---------|--------------------------------------------------------------------------|

> | `BUILDKIT_CACHE_MOUNT_NS`             | String  | Set optional cache ID namespace.                                         |

Reviewer: Do we have more details on these that we can link to ? (some example that describes how they're used)

Author: Added a simple example for one of them. Let me know if it lgty.

The `SHELL` instruction can also be used on Linux should an alternate shell be
required such as `zsh`, `csh`, `tcsh` and others.

## Here-Documents

> **Note**
>
> Added in [`docker/dockerfile:1.4`](#syntax)

Here-documents allow redirection of subsequent Dockerfile lines to the input of
`RUN` or `COPY` commands. If such command contains a [here-document](https://pubs.opengroup.org/onlinepubs/9699919799/utilities/V3_chap02.html#tag_18_07_04)

> Dockerfile will consider the next lines until the line only containing a

Reviewer 2: ```suggestion
the Dockerfile will consider the next lines until the line only containing a
```

Reviewer: Generally, avoid future tense in tech-docs; e.g. this could be

```markdown
the Dockerfile considers ...
```

here-doc delimiter as part of the same command.

### Example: Running a multi-line script

```dockerfile
# syntax=docker/dockerfile:1
FROM debian
RUN <<EOT bash
  apt-get update
  apt-get install -y vim
EOT
```

If the command only contains a here-document, its contents is evaluated with
the default shell.

```dockerfile
# syntax=docker/dockerfile:1
FROM debian
RUN <<EOT
  mkdir -p foo/bar
EOT
```

Alternatively, shebang header can be used to define an interpreter.

```dockerfile
# syntax=docker/dockerfile:1
FROM python:3.6
RUN <<EOT
#!/usr/bin/env python
print("hello world")
EOT
```

More complex examples may use multiple here-documents.

```dockerfile
# syntax=docker/dockerfile:1
FROM alpine
RUN <<FILE1 cat > file1 && <<FILE2 cat > file2
I am
first
FILE1
I am
second
FILE2
```

### Example: Creating inline files

In `COPY` commands source parameters can be replaced with here-doc indicators.
Regular here-doc [variable expansion and tab stripping rules](https://pubs.opengroup.org/onlinepubs/9699919799/utilities/V3_chap02.html#tag_18_07_04) apply.

```dockerfile
# syntax=docker/dockerfile:1
FROM alpine
ARG FOO=bar
COPY <<-EOT /app/foo

>   hello ${FOO}

Reviewer 2: This should be a tab to strip the whitespace at the front.
```suggestion
	hello ${FOO}
```

EOT
```

```dockerfile
# syntax=docker/dockerfile:1
FROM alpine
COPY <<-"EOT" /app/script.sh

>   echo hello ${FOO}

Reviewer 2: ^ same as above
